File: scripts/make_private.py
#!/usr/bin/env python3
"""
Cortex make-private script.
Flips your Cortex repo from public to private using a stored GitHub PAT.

Prerequisites:
  pip install cryptography
  python scripts/secrets.py store github-pat

Run:
  python scripts/make_private.py

The PAT needs the 'repo' scope on GitHub.
"""

import os
import sys
import json
import getpass
import argparse
import subprocess
import urllib.request
import urllib.error

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, os.path.join(ROOT, "scripts"))

try:
    from secrets import unified_get, perfile_names, LEGACY_VAULT
except ImportError as e:
    print(f"ERROR: Could not import from scripts/secrets.py: {e}")
    sys.exit(1)


def get_remote_url() -> str:
    result = subprocess.run(
        ["git", "remote", "get-url", "origin"],
        cwd=ROOT, capture_output=True, text=True
    )
    if result.returncode != 0:
        print("ERROR: Could not read git remote. Is this a git repo with an origin?")
        sys.exit(1)
    return result.stdout.strip()


def parse_owner_repo(remote_url: str) -> tuple:
    url = remote_url
    if url.startswith("git@"):
        # SSH: git@github.com:owner/repo.git or git@alias:owner/repo.git
        path = url.split(":", 1)[1]
    elif "github.com/" in url:
        # HTTPS: https://github.com/owner/repo.git
        path = url.split("github.com/", 1)[1]
    else:
        print(f"ERROR: Only GitHub remotes are supported. Got: {url}")
        sys.exit(1)
    path = path.removesuffix(".git")
    parts = path.split("/")
    if len(parts) != 2:
        print(f"ERROR: Could not parse owner/repo from: {url}")
        sys.exit(1)
    return parts[0], parts[1]


def print_manual_instructions(owner: str, repo: str) -> None:
    print()
    print("Cannot reach the GitHub API from this environment.")
    print("Expected on Claude mobile and other sandboxed agents — only git is permitted.")
    print()
    print(f"Flip {owner}/{repo} private manually:")
    print(f"  1. Open https://github.com/{owner}/{repo}/settings")
    print("  2. Scroll to the Danger Zone at the bottom")
    print("  3. Change repository visibility → Make private")
    print("  4. Confirm by typing the repo name")
    print()


def github_api_reachable() -> bool:
    # Preflight before prompting for the vault passphrase — on mobile/sandboxed agents
    # egress to api.github.com is blocked, and we should bail before the user types
    # their passphrase into chat for nothing.
    try:
        req = urllib.request.Request("https://api.github.com", method="HEAD")
        with urllib.request.urlopen(req, timeout=5):
            return True
    except urllib.error.URLError:
        return False


def main():
    parser = argparse.ArgumentParser(description="Flip Cortex repo to private via GitHub API")
    parser.add_argument("--passphrase", default=None, help="Vault passphrase (prompted if omitted)")
    args = parser.parse_args()

    remote_url = get_remote_url()
    owner, repo = parse_owner_repo(remote_url)
    print(f"Repo: {owner}/{repo}")

    if not github_api_reachable():
        print_manual_instructions(owner, repo)
        sys.exit(0)

    # Cheap pre-flight on v2 vault — spares the passphrase prompt when we can be
    # certain the secret isn't there. Legacy users (no per-file entry but legacy
    # blob present) fall through to unified_get, which decrypts the blob.
    if "github-pat" not in perfile_names() and not os.path.exists(LEGACY_VAULT):
        print("ERROR: No 'github-pat' in vault.")
        print("Store it first: python scripts/secrets.py store github-pat")
        sys.exit(1)

    passphrase = args.passphrase or getpass.getpass("Vault passphrase: ")
    pat = unified_get("github-pat", passphrase)

    payload = json.dumps({"private": True}).encode()
    req = urllib.request.Request(
        f"https://api.github.com/repos/{owner}/{repo}",
        data=payload,
        method="PATCH",
        headers={
            "Authorization": f"Bearer {pat}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
            "Content-Type": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(req) as resp:
            body = json.loads(resp.read())
            if body.get("private"):
                print(f"Done. {owner}/{repo} is now private.")
            else:
                print("WARNING: Request succeeded but repo does not appear private. Check GitHub.")
    except urllib.error.HTTPError as e:
        err = e.read().decode()
        print(f"ERROR: GitHub API {e.code}: {err}")
        sys.exit(1)
    except urllib.error.URLError as e:
        # Reached here despite preflight — network dropped mid-flight, or egress opened
        # up for the HEAD but not the PATCH. Don't leak a traceback.
        print(f"ERROR: Could not reach GitHub API: {e.reason}")
        print_manual_instructions(owner, repo)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: manifest/framework/scripts/make_private.ts
#!/usr/bin/env bun
/**
 * Cortex make-private script.
 * Flips your Cortex repo from public to private using a stored GitHub PAT.
 *
 * Prerequisites:
 *   bun manifest/framework/scripts/secrets.ts store github-pat
 *
 * Run:
 *   bun manifest/framework/scripts/make_private.ts [--passphrase <p>]
 *
 * The PAT needs the 'repo' scope on GitHub.
 */

import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import * as readline from 'node:readline'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..')

function parseArgs(argv: string[]): Record<string, string | undefined> {
  const result: Record<string, string | undefined> = {}
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]
    if (arg.startsWith('--')) {
      const key = arg.slice(2)
      const next = argv[i + 1]
      if (next && !next.startsWith('--')) {
        result[key] = next
        i++
      }
    }
  }
  return result
}

async function getRemoteUrl(): Promise<string> {
  const proc = Bun.spawn(['git', 'remote', 'get-url', 'origin'], {
    cwd: ROOT,
    stdout: 'pipe',
    stderr: 'pipe',
  })
  const code = await proc.exited
  if (code !== 0) {
    console.error('ERROR: Could not read git remote. Is this a git repo with an origin?')
    process.exit(1)
  }
  return (await new Response(proc.stdout).text()).trim()
}

function parseOwnerRepo(remoteUrl: string): [string, string] {
  let path: string
  if (remoteUrl.startsWith('git@')) {
    // SSH: git@github.com:owner/repo.git or git@alias:owner/repo.git
    path = remoteUrl.split(':').slice(1).join(':')
  } else if (remoteUrl.includes('github.com/')) {
    path = remoteUrl.split('github.com/')[1]
  } else {
    console.error(`ERROR: Only GitHub remotes are supported. Got: ${remoteUrl}`)
    process.exit(1)
  }
  path = path.replace(/\.git$/, '')
  const parts = path.split('/')
  if (parts.length !== 2) {
    console.error(`ERROR: Could not parse owner/repo from: ${remoteUrl}`)
    process.exit(1)
  }
  return [parts[0], parts[1]]
}

function printManualInstructions(owner: string, repo: string): void {
  console.log()
  console.log('Cannot reach the GitHub API from this environment.')
  console.log('Expected on Claude mobile and other sandboxed agents — only git is permitted.')
  console.log()
  console.log(`Flip ${owner}/${repo} private manually:`)
  console.log(`  1. Open https://github.com/${owner}/${repo}/settings`)
  console.log('  2. Scroll to the Danger Zone at the bottom')
  console.log('  3. Change repository visibility → Make private')
  console.log('  4. Confirm by typing the repo name')
  console.log()
}

async function githubApiReachable(): Promise<boolean> {
  try {
    const resp = await fetch('https://api.github.com', { method: 'HEAD', signal: AbortSignal.timeout(5000) })
    return resp.ok || resp.status < 500
  } catch {
    return false
  }
}

async function promptPassphrase(): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error('ERROR: No passphrase provided and stdin is not a TTY. Use --passphrase.')
    process.exit(1)
  }
  return new Promise(resolve => {
    process.stderr.write('Vault passphrase: ')
    let value = ''
    const tty = process.stdin as NodeJS.ReadStream & { setRawMode?: (mode: boolean) => void }
    if (tty.setRawMode) {
      process.stdin.setRawMode(true)
      process.stdin.resume()
      process.stdin.setEncoding('utf8')
      const onData = (ch: string) => {
        if (ch === '\r' || ch === '\n') {
          process.stdin.setRawMode!(false)
          process.stdin.pause()
          process.stdin.removeListener('data', onData)
          process.stderr.write('\n')
          resolve(value)
        } else if (ch === '') {
          process.exit(1)
        } else if (ch === '') {
          value = value.slice(0, -1)
        } else {
          value += ch
        }
      }
      process.stdin.on('data', onData)
    } else {
      const rl = readline.createInterface({ input: process.stdin, output: process.stderr })
      rl.question('', answer => {
        rl.close()
        resolve(answer)
      })
    }
  })
}

async function getSecret(name: string, passphrase: string): Promise<string> {
  const proc = Bun.spawn(
    ['bun', join(ROOT, 'manifest/framework/scripts/secrets.ts'), 'get', name, '--passphrase', passphrase],
    { stdout: 'pipe', stderr: 'pipe' }
  )
  const code = await proc.exited
  if (code !== 0) {
    const errText = await new Response(proc.stderr).text()
    console.error(`ERROR: Could not retrieve '${name}' from vault.`)
    console.error(errText.trim())
    process.exit(1)
  }
  return (await new Response(proc.stdout).text()).trim()
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2))

  const remoteUrl = await getRemoteUrl()
  const [owner, repo] = parseOwnerRepo(remoteUrl)
  console.log(`Repo: ${owner}/${repo}`)

  if (!(await githubApiReachable())) {
    printManualInstructions(owner, repo)
    process.exit(0)
  }

  const passphrase = args['passphrase'] ?? await promptPassphrase()
  const pat = await getSecret('github-pat', passphrase)

  try {
    const resp = await fetch(`https://api.github.com/repos/${owner}/${repo}`, {
      method: 'PATCH',
      headers: {
        Authorization: `Bearer ${pat}`,
        Accept: 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ private: true }),
    })

    if (!resp.ok) {
      const body = await resp.text()
      console.error(`ERROR: GitHub API ${resp.status}: ${body}`)
      process.exit(1)
    }

    const body = await resp.json() as { private?: boolean }
    if (body.private) {
      console.log(`Done. ${owner}/${repo} is now private.`)
    } else {
      console.log('WARNING: Request succeeded but repo does not appear private. Check GitHub.')
    }
  } catch (err) {
    console.error(`ERROR: Could not reach GitHub API: ${(err as Error).message}`)
    printManualInstructions(owner, repo)
    process.exit(1)
  }
}

main().catch(err => {
  console.error(`ERROR: ${err.message}`)
  process.exit(1)
})
